### ISSUE TITLE
Accept shortened but unambiguous words in BIP39 mnemonic

### ISSUE THREAD

**Issue by sowbug** [iancoleman/bip39#63]
When a user is typing in a mnemonic to recover from a backup, it's convenient to accept 4-character prefixes for BIP39 words, rather than requiring the user to type it all out.

Example: http://cryptosteel.com/ exploits the fact that the English BIP39 wordlist is unambiguous by the fourth character of each word, and is thus designed to preserve at most four characters of each word. Thus, "BROC" would be interpreted as "BROCCOLI."

**Comment by iancoleman** [comment]
Thanks for the suggestion.

This is a good idea and I agree some sort of autocomplete functionality would be useful.

When I get a chance I'll look into implementing this, since it makes it easier for users to enter their mnemonic.

My initial feeling is to always show mnemonic phrases with the full words rather than the shortened version, eg show 'ankle task mandate jungle tornado' but not 'ankl task mand jung torn'... but we'll see how this feature unfolds from a UI perspective.

**Comment by sowbug** [comment]
I agree about showing the full words. There is virtuous redundancy in presenting all the characters, which will help in cases where the user misspells a word, or part of the printed medium is damaged.

**Comment by erikbos** [comment]
I have entered my words a few times in the mnemonic converter and practiced doing it from memory at the same time.. I realised I remember actually quite a lot of words but make small mistakes like spill vs split. Most ideal would imho autocompletion to the full word to aid memorising the correct words.

**Comment by iancoleman** [comment]
I tried implementing this today as an autosuggest type feature. It used `setSelectionRange` on the autosuggested / untyped portion of the final word. But it was not a good user experience; it (unsurprisingly) got really messy trying to alter the user input while they typed.

For example, editing a word in the middle of the mnemonic would not show the autosuggestion. There were edge cases when pressing backspace or delete. Just lots of messy stuff.

I think a better method will be to let the user type `brai refu moti` but then show a read-only field below that with the complete and tidy mnemonic `brain refuse motion`.

**Comment by sowbug** [comment]
Thanks for prototyping it! Now that I think back to Code Editors I've Known And Loved, the autocomplete feature is indeed more subtle and complex than it feels when using it. I wonder whether it'd be helpful to try dropping in any of the jQuery autocomplete libraries out there ([here's one](https://jqueryui.com/autocomplete/#multiple)) to find a UX that works for this use case, and then either include it after a security review, or write a smaller version of the same UX.
